pilot/pkg/networking/core/v1alpha3/tls.go

```diff
@@ -0,0 +1,140 @@
+	for _, config := range configs {
+		virtualService := config.Spec.(*v1alpha3.VirtualService)
+		for _, vsHost := range virtualService.Hosts {
+			if model.Hostname(vsHost).Matches(host) {
+				return virtualService
+			}
+		}
+	}
+	return nil
+}
+
+// TODO: check port protocol?
+func buildOutboundTCPFilterChainOpts(proxy model.Proxy, env model.Environment, configs []model.Config, addresses []string,
+	host model.Hostname, listenPort *model.Port, proxyLabels model.LabelsCollection, gateways map[string]bool) []*filterChainOpts {
+
+	out := make([]*filterChainOpts, 0)
+
+	virtualService := getVirtualServiceForHost(host, configs)
+	if virtualService != nil {
+		// TODO: Make SNI compatible with RBAC. Deprecated tcp route configs are incompatible with SNI.
+		// RBAC requires deprecated tcp route configs, so RBAC is incompatible with SNI.
+		for _, tls := range virtualService.Tls {
+			// since we don't support weighted destinations yet there can only be exactly 1 destination
+			dest := tls.Route[0].Destination
+			destSvc, err := env.GetService(model.Hostname(dest.Host))
+			if err != nil {
+				log.Debugf("failed to retrieve service for destination %q: %v", host, err)
+				continue
+			}
+			clusterName := istio_route.GetDestinationCluster(dest, destSvc, listenPort.Port)
+			for _, match := range tls.Match {
+				if sourceMatchTLS(match, proxyLabels, gateways, listenPort.Port) {
+					out = append(out, &filterChainOpts{
+						sniHosts:       match.SniHosts,
+						networkFilters: buildOutboundNetworkFilters(clusterName, nil, listenPort),
+					})
+				}
+			}
+		}
+
+		// very basic TCP (no L4 matching)
+		// break as soon as we add one network filter
+	TcpLoop:
+		for _, tcp := range virtualService.Tcp {
+			// since we don't support weighted destinations yet there can only be exactly 1 destination
+			dest := tcp.Route[0].Destination
+			destSvc, err := env.GetService(model.Hostname(dest.Host))
+			if err != nil {
+				log.Debugf("failed to retrieve service for destination %q: %v", host, err)
+				continue
+			}
+			clusterName := istio_route.GetDestinationCluster(dest, destSvc, listenPort.Port)
+			addresses = []string{destSvc.GetServiceAddressForProxy(&proxy)}
+
+			if len(tcp.Match) == 0 { // implicit match
+				out = []*filterChainOpts{{
+					networkFilters: buildOutboundNetworkFilters(clusterName, addresses, listenPort),
+				}}
+				break TcpLoop
+			}
+			for _, match := range tcp.Match {
+				if sourceMatchTCP(match, proxyLabels, gateways, listenPort.Port) {
+					out = []*filterChainOpts{{
+						networkFilters: buildOutboundNetworkFilters(clusterName, addresses, listenPort),
+					}}
+					break TcpLoop
+				}
+			}
+		}
+	}
+
+	if len(out) == 0 { // FIXME:
+		clusterName := model.BuildSubsetKey(model.TrafficDirectionOutbound, "", host, int(listenPort.Port))
+		out = []*filterChainOpts{{
+			networkFilters: buildOutboundNetworkFilters(clusterName, addresses, listenPort),
+		}}
```

Comment:
this is incorrect. For TLS, we need to have virtualservices to route.. Otherwise, the traffic goes no where.. But for TCP, we need default routes for every service in the service registry. So, you might need some marker, like portname to gate this default cluster creation.